File: src/core/git_operations.cpp
#include "aliases/git_operations.h"
#include "aliases/common.h"
#include "aliases/file_utils.h"
#include "aliases/process_utils.h"

namespace aliases {

GitStatus GitOperations::get_git_status(const std::string& directory) {
    GitStatus status;

    // Check if it's a git repository
    auto git_dir = FileUtils::join_path(directory, ".git");
    status.is_git_repo = FileUtils::directory_exists(git_dir);

    if (!status.is_git_repo) {
        return status;
    }

    // Get current branch
    status.current_branch = get_current_branch(directory);
    status.is_main_branch = is_main_branch(status.current_branch);

    // Check for uncommitted changes
    status.has_uncommitted_changes = has_uncommitted_changes(directory);

    return status;
}

std::string GitOperations::get_current_branch(const std::string& directory) {
    auto result = ProcessUtils::execute("git rev-parse --abbrev-ref HEAD", directory);
    if (result.success()) {
        return trim(result.stdout_output);
    }
    return "";
}

bool GitOperations::has_uncommitted_changes(const std::string& directory) {
    auto result = ProcessUtils::execute("git status --porcelain", directory);
    return result.success() && !trim(result.stdout_output).empty();
}

bool GitOperations::is_git_repository(const std::string& directory) {
    auto git_dir = FileUtils::join_path(directory, ".git");
    return FileUtils::directory_exists(git_dir);
}

Result<std::string> GitOperations::checkout_branch(const std::string& directory, const std::string& branch) {
    auto result = ProcessUtils::execute("git checkout " + branch, directory);
    if (result.success()) {
        return Result<std::string>::success_with(std::string("Switched to branch " + branch));
    } else {
        return Result<std::string>::error("Failed to checkout branch: " + result.stderr_output);
    }
}

Result<std::string> GitOperations::pull_changes(const std::string& directory) {
    auto result = ProcessUtils::execute("git pull --ff-only", directory);
    if (result.success()) {
        return Result<std::string>::success_with(std::move(result.stdout_output));
    } else {
        return Result<std::string>::error("Failed to pull changes: " + result.stderr_output);
    }
}

bool GitOperations::is_main_branch(const std::string& branch_name) {
    return branch_name == "main" || branch_name == "master";
}

std::string GitOperations::get_main_branch_name(const std::string& directory) {
    // Try to determine main branch name
    auto result = ProcessUtils::execute("git symbolic-ref refs/remotes/origin/HEAD", directory);
    if (result.success()) {
        auto output = trim(result.stdout_output);
        auto pos = output.find_last_of('/');
        if (pos != std::string::npos) {
            return output.substr(pos + 1);
        }
    }

    // Default fallback
    return "main";
}

} // namespace aliases

File: aliases/git_operations.py
"""Thin wrappers around git CLI commands."""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Union

from aliases.process_utils import execute

MAIN_BRANCH_NAMES = frozenset({"main", "master", "trunk", "develop"})


@dataclass
class GitStatus:
    is_git_repo: bool
    current_branch: str = ""
    has_uncommitted_changes: bool = False
    is_main_branch: bool = False


def get_status(path: Union[Path, str]) -> GitStatus:
    """Return the git status for the repo at *path*."""
    code, _, _ = execute(["git", "rev-parse", "--git-dir"], cwd=path)
    if code != 0:
        return GitStatus(is_git_repo=False)

    _, branch_out, _ = execute(["git", "rev-parse", "--abbrev-ref", "HEAD"], cwd=path)
    branch = branch_out.strip()

    _, status_out, _ = execute(["git", "status", "--porcelain"], cwd=path)

    return GitStatus(
        is_git_repo=True,
        current_branch=branch,
        has_uncommitted_changes=bool(status_out.strip()),
        is_main_branch=branch.lower() in MAIN_BRANCH_NAMES,
    )


def get_current_branch(path: Union[Path, str]) -> str | None:
    code, out, _ = execute(["git", "rev-parse", "--abbrev-ref", "HEAD"], cwd=path)
    return out.strip() if code == 0 else None


def is_git_repo(path: Union[Path, str]) -> bool:
    code, _, _ = execute(["git", "rev-parse", "--git-dir"], cwd=path)
    return code == 0


def pull_changes(path: Union[Path, str]) -> tuple[bool, str]:
    """Run ``git pull`` in *path*. Returns ``(success, output)``."""
    code, out, err = execute(["git", "pull"], cwd=path)
    return code == 0, (out + err).strip()


def is_main_branch(branch: str) -> bool:
    return branch.lower() in MAIN_BRANCH_NAMES
